h2o-core/src/main/java/water/parser/PreviewParseWriter.java

```diff
@@ -131,86 +131,102 @@ public void addNAs(int colIdx, int nrow) {
-      if(_nstrings[i] >= (_nnums[i]+_nzeros[i]) // mostly strings
-              && (_domains[i].size() <= 0.95 * _nstrings[i]) ) { // but not all unique
-        types[i] = Vec.T_CAT;
-        continue;
-      }
-
-      // All guesses failed
-      types[i] = Vec.T_NUM;
+      IcedHashMap<String, String> sourceDomain = _domains[i];
+      IDomain domain = new IDomain() {
+        public int size() { return sourceDomain.size(); }
+        public boolean contains(String value) { return sourceDomain.containsKey(value); }
+      };
+      types[i] = PreviewParseWriter.guessType(
+        _nlines,
+        _nnums[i],
+        _nstrings[i],
+        _ndates[i],
+        _nUUID[i],
+        _nzeros[i],
+        _nempty[i],
+        domain);
     }
     return types;
   }
+  
+  public interface IDomain {
+    int size();
+    boolean contains(String value);
+  }
+
+  public static byte guessType(
+      int nlines,
+      int nnums,
+      int nstrings,
+      int ndates,
+      int nUUID,
+      int nzeros,
+      int nempty,
+      IDomain domain) {
+    int nonemptyLines = nlines - nempty - 1; //During guess, some columns may be shorted one line based on 4M boundary
+    
+    //Very redundant tests, but clearer and not speed critical
+    
+    // is it clearly numeric?
+    if ((nnums + nzeros) >= ndates
+        && (nnums + nzeros) >= nUUID
+        && nnums >= nstrings) { // 0s can be an NA among categoricals, ignore
+          return Vec.T_NUM;
+    }
+    
+    // All same string or empty?
+    if (domain.size() == 1 && ndates==0 ) {
+      // Obvious NA, or few instances of the single string, declare numeric
+      // else categorical
+      return (domain.contains("NA") ||
+              domain.contains("na") ||
+              domain.contains("Na") ||
+              domain.contains("N/A") ||
+              nstrings < nnums+nzeros) ? Vec.T_NUM : Vec.T_CAT;
+    }
+    
+    // with NA, but likely numeric
+    if (domain.size() <= 1 && (nnums + nzeros) > ndates + nUUID) {
+      return Vec.T_NUM;
+    }
+    
+    // Datetime
+    if (ndates > nUUID && ndates > (nnums + nzeros) && (ndates > nstrings || domain.size() <= 1)) {
+      return Vec.T_TIME;
+    }
+    
+    // UUID
+    if (nUUID > ndates && nUUID > (nnums + nzeros) && (nUUID > nstrings || domain.size() <= 1)) {
+      return Vec.T_UUID;
+    }
+    
+    // Strings, almost no dups
+    if (nstrings > ndates
+        && nstrings > nUUID
+        && nstrings > (nnums + nzeros)
+        && domain.size() >= 0.95 * nstrings) {
+      return Vec.T_STR;
+    }
+    
+    // categorical or string?
+    // categorical with 0s for NAs
+    if (nzeros > 0
+        && ((nzeros + nstrings) >= nonemptyLines) //just strings and zeros for NA (thus no empty lines)
+        && (domain.size() <= 0.95 * nstrings)) { // not all unique strings
+      return Vec.T_CAT;
+    }
+    // categorical mixed with numbers
+    if (nstrings >= (nnums+nzeros) // mostly strings
```

Comment:
Unnecessary parentheses + wrong formatting (minor issue). Feel free to keep the parentheses in place if you feel like it, please. This is only a matter of personal taste.

Reply:
@Pscheidl It was like this before, I just changed the variable names. I will fix the parentheses and spaces around `+`, but will leave it splitted the condition into two lines due to preserving the comments.

Reply:
On second thought, I've decided to leave the parentheses as they are. They are used like this on more places within this class. If the usage is inappropriate, we could fix it in a different PR, just dedicated to this problem.